### PR TITLE
MGMT-4869 add an assisted operator installation workflow

### PR DESCRIPTION
**Pull request by osherdp**
/hold
/cc @osherdp

### CODE REVIEW

File: ci-operator/step-registry/baremetalds/assisted/operator/lso/install/baremetalds-assisted-operator-lso-install-commands.sh [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+spec:
+  channel: preview
+  name: local-storage-operator
+  source: localstorage-operator-manifests
+  sourceNamespace: openshift-local-storage
+EOCR
+
+echo "Waiting for LocalVolume CRD to be defined..."
+for i in {1..60}; do
+    oc get crd/localvolumes.local.storage.openshift.io -n openshift-local-storage && break || sleep 10
```
Comment (YuviGold) [inline ci-operator/step-registry/baremetalds/assisted/operator/lso/install/baremetalds-assisted-operator-lso-install-commands.sh:56]:
```bash
cat << EOF | tee "lso_crds.yaml"
...
EOF

oc create -f "lso_crds.yaml"
oc wait --for=condition=Established -f "lso_crds.yaml"
```

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  namespace: openshift-marketplace
+spec:
+  sourceType: grpc
+  image: quay.io/ocpmetal/assisted-service-index:latest
+EOCR
+
+SECRET=\$(oc get secret --field-selector type=kubernetes.io/dockercfg -o name | grep default)
+PULL_SECRET=\$(oc get \$SECRET -o json | jq -r '.data[".dockercfg"]' | base64 -d)
+
+echo "Creating pull-secret..."
```
Comment (YuviGold) [inline ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:95]: used by who?

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  installPlanApproval: Automatic
+  name: hive-operator
+  source: community-operators
+  sourceNamespace: openshift-marketplace
+  startingCSV: hive-operator.v1.0.19
+EOCR
+
+wait_for_crd "clusterdeployments.hive.openshift.io"
+
+echo "Install PVCs and other prerequisites for assisted-installer operator..."
```
Comment (YuviGold) [inline ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:43]: I think that the PVC should be defined after the OLM prerequisites

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  fi
+
+  sleep 10
+done
+
+echo "all resources:"
+oc get all -n assisted-installer
+
+echo "waiting for assisted-service pod to be ready..."
+oc wait -n assisted-installer --for=condition=Ready pod -l app=assisted-service --timeout=90s
```
Comment (YuviGold) [inline ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:175]: What's the difference between the two waiting?

File: ci-operator/step-registry/baremetalds/assisted/operator/lso/create-disks/baremetalds-assisted-operator-lso-create-disks-commands.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+# Fetch packet basic configuration
+# shellcheck source=/dev/null
+source "${SHARED_DIR}/packet-conf.sh"
+
+ssh "${SSHOPTS[@]}" "root@${IP}" bash - << EOF |& sed -e 's/.*auths\{0,1\}".*/*** PULL_SECRET ***/g'
+
+set -xeo pipefail
+
+echo "Creating disks..."
+for node in $(virsh list --name | grep ostest_worker); do
```
[the remark this answers is not in the document]
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/lso/create-disks/baremetalds-assisted-operator-lso-create-disks-commands.sh:18]: niiice...

File: ci-operator/step-registry/baremetalds/assisted/operator/lso/install/baremetalds-assisted-operator-lso-install-commands.sh (hunk 1 above)
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/lso/install/baremetalds-assisted-operator-lso-install-commands.sh:56]: interesting
you say that ``oc create`` will only return after the creation? in oppose to ``oc apply``

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh (hunk 2 above)
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:95]: by the service. No?

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh (hunk 4 above)
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:175]: hmm, waiting for creation of the object and waiting for the pod to become ready
but maybe I'll replace with waiting for the CSV using the code you sent to me

File: ci-operator/step-registry/baremetalds/assisted/operator/baremetalds-assisted-operator-workflow.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  as: baremetalds-assisted-operator
+  steps:
+    pre:
+    - ref: baremetalds-packet-setup
+    - ref: ipi-install-rbac
+    - ref: baremetalds-devscripts-setup
+    - chain: baremetalds-assisted-operator-lso
+    - ref: baremetalds-assisted-operator-setup
+    post:
+    # - ref: baremetalds-assisted-gather
```
Comment (osherdp) [inline ci-operator/step-registry/baremetalds/assisted/operator/baremetalds-assisted-operator-workflow.yaml:11]: remove

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh (hunk 3 above)
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:65]: removed PVCs creation here
I'll fix the ``echo`` above as well

File: ci-operator/step-registry/baremetalds/assisted/operator/baremetalds-assisted-operator-workflow.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    - ref: baremetalds-devscripts-setup
+    - chain: baremetalds-assisted-operator-lso
+    - ref: baremetalds-assisted-operator-setup
+    post:
+    - ref: baremetalds-devscripts-gather
+    - ref: baremetalds-packet-teardown
+    env:
+      DEVSCRIPTS_CONFIG: |
+        IP_STACK=v4
+        NUM_WORKERS=4
```
Comment (lranjbar) [inline ci-operator/step-registry/baremetalds/assisted/operator/baremetalds-assisted-operator-workflow.yaml:16]: Do we need 4 workers? Default set up is NUM_WORKERS=2 this should be sufficient for now

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh (hunk 3 above)
Comment (lranjbar) [reply ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh:65]: We can probably do it in this order:
- Install operators
- Create local storage using LSO
- Create PVCs
- Run assisted-installer operand from assisted-installer operator

I think that should solve the order of operations for the 3 operators. That might be the current problem from what I'm seeing from the failures.

File: ci-operator/step-registry/baremetalds/assisted/operator/baremetalds-assisted-operator-workflow.yaml (hunk 7 above)
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/baremetalds-assisted-operator-workflow.yaml:16]: yes, I think you're right

File: ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh (hunk 3 above)
Comment (osherdp) [reply ci-operator/step-registry/baremetalds/assisted/operator/setup/baremetalds-assisted-operator-setup-commands.sh]: no dependency problems, for the best of my knowledge

**Comment by osherdp** [comment]
/unhold
before @YuviGold changes him mind :)

**Pull request merged**